### PR TITLE
Fix failures exposed via test_datetime.py

### PR DESCRIPTION
**Pull request by rassilon**
* constructing a subclass of datetime/date failed due to the protected constructor
* pickling a subclass of datetime/date/time faile.
* time was missing its __format__ implementation.
  test_datetime tests overloading strftime, so __format__ calls it if necessary.
* Apply the same pickling fix to SetCollection/FrozenSetCollection and deque.

This PR has fixes for these problems.

### CODE REVIEW

File: Languages/IronPython/IronPython.Modules/datetime.cs [hunk 1]
```diff
@@ -1486,6 +1494,22 @@ public class time : ICodeFormattable {
...
 
+            public object __format__(CodeContext/*!*/ context, string dateFormat) {
+                if (string.IsNullOrEmpty(dateFormat)) {
+                    return PythonOps.ToString(context, this);
+                }
+                else {
+                    // If we're a subtype, there might be a strftime overload,
+                    // so call it if it exists.
+                    if (GetType() == typeof(time)) {
+                        return this.strftime(context, dateFormat);
```
Comment (slide) [inline Languages/IronPython/IronPython.Modules/datetime.cs]: No need to use "this."

Comment (rassilon) [inline Languages/IronPython/IronPython.Modules/datetime.cs]: That was a copy/paste from date.__format__, but I've cleaned up both instances in the following commit.

**Comment by slide** [comment]
LGTM

**Comment by slide** [comment]
@jdhardy Any issues with this?

**Comment by jdhardy** [comment]
LGTM

**Pull request merged**